man/vzcptcheck.8.in
.TH vzcptcheck 8 "21 Apr 2015" "OpenVZ" "Containers"
.SH NAME
vzcptcheck \- show/test CPT properties
.SH SYNOPSIS
.SY vzcptcheck
.I property
.OP argument ...
.SY vzcptcheck
.B help
.YS
.SH DESCRIPTION
This low-level utility is used by
.BR vzmigrate (8)
to check if a container can be succesfully restored on a destination system
before performing live migration. Currently, it is able to check CPT version
and CPU flags compatibility.
.P
There are two modes of operation: to show a property and to test a
property. Showing mode is invoked if a property specified with no arguments;
testing mode requires one or many arguments after a property name.
Generally, one runs
.B vzcptcheck
on one node to get a property, and on the other node to check for the
property.
.P
Note that this utility does not guarantee that migration will succeed
even if all tests are passed. It merely gives a way to prevent migration
which will definitely fail.
.SH OPTIONS
.SY vzcptcheck
.B version
.SY vzcptcheck
.B version
.I num
.YS
To get CPT version number, run \fBvzcptcheck version\fR on the source node.
Then, to check if this version number is supported, run
\fBvzcptcheck version \fInum\fR on the destination node.
.SY vzcptcheck
.B caps
.SY vzcptcheck
.B caps
.I ctid
.I capnum
.SY vzcptcheck
.B caps
.I capnum
.YS
To get CPU capabilities number, run \fBvzcptcheck caps\fR on the destination
node. Then, to check if a running CT can be live migrated to a node
with such a CPU, run \fBvzcptcheck caps \fIctid capnum\fR on the source node.
To check if a suspended container can be migrated (i.e. it can be restored
on the destination node), use the form without \fIctid\fR:
\fBvzcptcheck caps \fIcapnum\fR.
.SH EXIT STATUS
In showing mode, returns \fB0\fR upon success, or \fB1\fR in case of an error.
.P
In testing mode, returns \fB0\fR if migration is possible (test passed),
\fB1\fR in case of an error, or \fB2\fR if migration is not possible
(test failed).
.SH EXAMPLES
To check if a CT 123 can be migrated to $DEST, run the following commands
on the source node. Note that error checking is omitted for brevity.
.P
.EX
   VEID=123
   VERSION=$(vzcptcheck version)
   echo Got CPT version: $VERSION
   ssh root@$DEST vzcptcheck version $VERSION
   echo Version check result: $?
   CAPS=$(ssh root@$DEST vzcptcheck caps)
   echo Got caps: $CAPS
   vzcptcheck caps $VEID $CAPS
   echo Caps check result: $?
.EE
.P
If both tests pass, you should get 0 as a result for both checks.
.SH SEE ALSO
.BR vzmigrate (8).
.SH LICENSE
Copyright (C) 2013-2015, Parallels, Inc. Licensed under GNU GPL v2.


ChangeLog
Since after 3.0.12 release, this file is obsoleted. Complete change log
can be obtained from OpenVZ git repository.

	http://git.openvz.org/?p=vzctl;a=shortlog
	http://git.openvz.org/?p=vzctl;a=log

Below is a change log for old vzctl releases.


   3.0.12: Thu Oct 5 2006 <igor@openvz.org>
	* vzctl start: added --wait cmdline option
	* added vznetcfg script for veth setup
	* fixed error message in case kernel compiled without USER_RESOURCES
	* added '--applyconfig_map name' vzctl set option
	* changed friendly name data structure
	* fixed vzmigrate to support VE name feature (bug #239)
	* fixed annoying arp warning (bug #189)
	* added bash completion script (bug #212)
	* added check to vzmigrate if OpenVZ is run on another side (bug #163)
	* removed debian directory from the tarball (bugs #197, #194)
	* removed LICVIEW from /etc/init.d/vz (bug #237)
	* fixed misprints and /dev/nul(l) usage in dist scripts (bug #228)
	* fixed vzmigrate to work with different awk flavors (bug #231)
	* years in copyright fixed (bug #232)
	* increased some UBCs in ve-light.conf-sample (bug #234)
	* fixed some minor memory leaks while doing realloc()
	* vzctl destroy: remove VE names on destroy (bug #236)
	* added --name option description to vzctl(8) man page (bug #232)
	* merged libvzctl-simfs and libvzctl into a single libvzctl
	* vzlist: replaced sprintf() with snprintf() to avoid buffer overflows
	* vzctl enter: forward error messages from child process
	* vzctl enter: fix off-by-one error which resulted in
	  uninitialized PATH
	* arpsend: fixed segfault when parsing long options
	* vzlist: fixed cast from pointer to integer of different size warnings
	* vzctl enter: exit with correct error code on enter fail (bug #245)
	* added udev rules for /dev/vzctl (bug #204)
	* added syscall numbers for powerpc
	* /etc/init.d/vz: check for /dev/vzctl only after modprobe (bug #204)
	* fixed ppc64 compilation
	* removed distro-specific stuff from 'make install' (bug #254)
	* added 'make install-gentoo' and Gentoo init.d/vz script (bug #262)
	* fixed arpsend options parsing for some compilers (bug #261)
	* man pages cleanup (VPS->VE, etc.)
	* changed all utilities and man pages to refer to VE not VPS
	* vzcpucheck: give error if VE0 CPU units not set (bug #249)

   3.0.11: Mon Aug  7 2006 <igor@openvz.org>
	* fixed using VE_ROUTE_SRC_DEV variable
	* added VPS name support
	* added change number of CPUs available in the VPS functionality.
	* fixes in CPT error processing
	* added upport of xt_mac iptables module
	* fixed awk expression to be compatible with mawk

   3.0.10: Wed Jun  7 2006 <igor@openvz.org>
	* added virtual ethernet device support (--veth_add, --veth_del)
	* added  /proc/meminfo virtualization support (--meminfo mode:value)
	* added VE0 UB setup functionality (--reset-ub)

   3.0.9: Wed May 24 2006 <igor@openvz.org>
	vzctl:
	* switched to new versioning scheme.
	* added -z option to /sbin/init
		(http://bugzilla.openvz.org/show_bug.cgi?id=164)
	* reset standart file desc [0-2] to /dev/null in case they are closed.
	* added possibility to set cpulimit in %
	* added suffixes G,M,K support for --diskspace option
	* set SIGPIPE handler to SIG_IGN before fist write to stdout/err
	* added Debian packaging support
	* changed directory structure.
		/etc/sysconfig/vz-scripts -> /etc/vz/conf
		/etc/sysconfig/vz-scripts/dists -> /etc/vz/dists
		/etc/sysconfig/vz -> /etc/vz/vz.conf
	* better distribution detection in vz initscript.
	* added Ubuntu distribution configuration file as copy of debian.
		(http://bugzilla.openvz.org/show_bug.cgi?id=119)
	* Increased kmemsize, privmpages, numfiles for x86_64
		ve-vps.basic.conf-sample.
		(http://bugzilla.openvz.org/show_bug.cgi?id=69)
	* changed -p -> --preserve=mode,ownership for 'cp' workaround in
		case /proc is not mounted.
	* internal modifications
		changed dynamic UB structure to static one.
		moved VPS configure functionality to vps_configure.c
		moved destroy functionaliny to destroy.c
	vzmigrate:
	* move config as veid.migrated to allow backward migration
		(http://bugzilla.openvz.org/show_bug.cgi?id=152)
	* added coping of action scripts.
		(http://bugzilla.openvz.org/show_bug.cgi?id=157)
	* fixed --keep-dst option processing. Option behavior was inverted.

   3.0.0-8: Mon Apr 24 2006 <igor@openvz.org>
	* UBC parameters > LONG_MAX in VPS conf truncate to LONG_MAX
	* added installation of vzmigrate.8 man page
	* use 'uname -m' for arch selection

   3.0.0-6: Tue Apr 18 2006 <kir@openvz.org>
	* man page fixes
   3.0.0-5: Fri Apr  7 2006 <igor@sw.ru>
	* added checkpointing support
	* added kernel headers for standalone compilation
	* chaged license from QPL to GPL
	* set personality PER_LINUX32 on x86_64 arch for i386 VPSes
	* fixed --iptables store (do not store dependent modules)
	* Do skip unknown iptables modules in config
	* set arg[0] = "-bash" on enter

   3.0.0-4: Thu Mar 16 2006 <kir@openvz.org>
	* cron scripts do nothing if VZ is not loaded (bug #107)
	* cosmetic vzctl.8 man page fixes
	* reverted bug #110 changes, see http://bugzilla.openvz.org/111
	* added ipt_REDIRECT support, thanks to jstubbs-at-gentoo.org

   3.0.0-3: Thu Mar  2 2006 <igor@sw.ru>
	* put_param3() can incorrectly set $name=( "" ) (bug #111).
		Affects gentoo VPSs. Thanks to jstubbs-at-gentoo.org
	* fixed compilation against old kernels (<= 022stab070)
	* description of set parameter suffixes (K, M, G, P) added
		to vzctl.8 man page (bug #105)
	* fixed gentoo-add_ip.sh script (bug #110)

   3.0.0-2: Thu Feb 23 2006 <igor@sw.ru>
	* fixed VPS start on old kernel (witthout feature_mask support)

   3.0.0-1: Wed Feb 22 2006 <igor@sw.ru>
	* enable sysfs for SuSE like distributions
	* added possibility to setup UBC parameters in the form
		xxx[GMKPB]:yyy[GMKPB]

   2.7.0-27: Tue Feb  7 2006 <igor@sw.ru>
	* fixed VE_CONFFILE variable passed to vps.mount script
	* fixed venet0 initialization in vz initscript (bug #99)

   2.7.0-26: Thu Feb  2 2006 <igor@sw.ru>

src/vzcptcheck.c
/*
 *  Copyright (C) 2013-2015, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <limits.h>
#include <errno.h>
#include <sys/ioctl.h>

#include <linux/cpt_ioctl.h>

static int usage(int rc)
{
	FILE *out = (rc == 0)? stdout : stderr;

	fprintf(out,
"Usage:	vzcptcheck <command> [argument ...]\n"
" [S]	vzcptcheck version		show cpt in-kernel version\n"
" [D]	vzcptcheck version <ver>	test if <ver> version is supported\n"
" [D]	vzcptcheck caps			show CPU capabilities\n"
" [S]	vzcptcheck caps <caps>		test if <caps> work\n"
" [S]	vzcptcheck caps <veid> <caps>	test if <caps> work for running <veid>\n"
"\n"
"[S] -- run this on source node\n"
"[D] -- run on destination node\n"
	);

	return rc;
}

static void ioctl_error(const char *str)
{
	fprintf(stderr, "Error in ioctl(%s): %m\n", str);
}

static int cpt_open(const char *f, int *error_fd)
{
	char file[256];
	int fd;
	int error_pipe[2];

	if (error_fd != NULL) {
		/* Prepare CPT error fd */
		if (pipe(error_pipe) < 0) {
			fprintf(stderr, "Can't create pipe: %m\n");
			return -1;
		}
		fcntl(error_pipe[0], F_SETFL, O_NONBLOCK);
		fcntl(error_pipe[1], F_SETFL, O_NONBLOCK);
	}

	/* Open the proc file */
	snprintf(file, sizeof(file), "/proc/%s", f);
	fd = open(file, O_RDWR);
	if (fd < 0) {
		fprintf(stderr, "Can't open %s: %m\n", file);
		fprintf(stderr, "Kernel module vz%s not loaded\n", f);
		return -1;
	}

	if (error_fd != NULL) {
		/* Setup CPT error fd */
		if (ioctl(fd, CPT_SET_ERRORFD, error_pipe[1]) < 0) {
			ioctl_error("CPT_SET_ERRORFD");
			close(fd);
			return -1;
		}
		close(error_pipe[1]);
		*error_fd = error_pipe[0];
	}

	return fd;
}

static void cpt_read_error(int fd)
{
	int len, len1;
	char buf[PIPE_BUF];

	while ((len = read(fd, buf, PIPE_BUF)) > 0)
		do {
			len1 = write(STDERR_FILENO, buf, len);
			len -= len1;
		} while (len > 0 && len1 > 0);
}

static int cpt_version(int argc, char *argv[])
{
	int fd, version;

	if (argc == 1) { /* show version */
		fd = cpt_open("cpt", NULL);
		if (fd < 0)
			return 1;

		version = ioctl(fd, CPT_TEST_VERSION);
		close(fd);
		if (version <= 0) {
			ioctl_error("CPT_TEST_VERSION");
			return 1;
		}

		printf("%d\n", version);
		return 0;
	}
	else if (argc == 2) { /* test version */
		int ret;

		version = atoi(argv[1]);
		fd = cpt_open("rst", NULL);
		if (fd < 0)
			return 1;

		ret = ioctl(fd, CPT_TEST_VERSION, version);
		close(fd);
		if (ret < 0) {
			if (errno == EINVAL) /* no such ioctl -- can migrate */
				return 0;
			ioctl_error("CPT_TEST_VERSION");
			return 1;
		}
		else if (ret == 1) /* version accepted */
			return 0;

		/* else -- version rejected */
		fprintf(stderr, "CPT version %d rejected: can't migrate\n",
				version);

		return 2;

	}

	/* else */
	return usage(1);
}

static int cpt_caps(int argc, char *argv[])
{
	int fd, err_fd, caps;

	if (argc == 1) { /* show caps */
		fd = cpt_open("rst", &err_fd);
		if (fd < 0)
			return 1;

		caps = ioctl(fd, CPT_TEST_CAPS);
		cpt_read_error(err_fd);
		close(fd);
		if (caps <= 0) {
			ioctl_error("CPT_TEST_CAPS");
			return 1;
		}

		printf("%d\n", caps);
		return 0;
	}
	else if (argc == 2) { /* test caps, no VEID */
		int ret;

		caps = atoi(argv[1]);

		fd = cpt_open("cpt", &err_fd);
		if (fd < 0)
			return 1;

		ret = ioctl(fd, CPT_TEST_CAPS, caps);
		cpt_read_error(err_fd);
		close(fd);
		if (ret < 0) {
			if (errno == ENOSYS || errno == EAGAIN) {
				fprintf(stderr, "vzcptcheck WARNING: old "
						"kernel, please update to "
						"kernel >= 042stab107.x\n");
				/* As we can't test caps anyway, assume OK */
				return 0;
			}
			ioctl_error("CPT_TEST_CAPS");
			return 1;
		}

		if (ret > 0) {
			fprintf(stderr, "Insufficient CPU capabilities: "
					"can't migrate\n");
			return 2;
		}

		return 0;
	}
	else if (argc == 3) { /* test caps */
		int veid, ret;

		veid = atoi(argv[1]);
		caps = atoi(argv[2]);

		fd = cpt_open("cpt", &err_fd);
		if (fd < 0)
			return 1;

		ret = ioctl(fd, CPT_SET_VEID, veid);
		if (ret < 0) {
			ioctl_error("CPT_SET_VEID");
			close(fd);
			return 1;
		}

		ret = ioctl(fd, CPT_TEST_VECAPS, caps);
		cpt_read_error(err_fd);
		close(fd);
		if (ret < 0) {
			ioctl_error("CPT_TEST_VECAPS");
			return 1;
		}

		if (ret > 0) {
			fprintf(stderr, "Insufficient CPU capabilities: "
					"can't migrate\n");
			return 2;
		}

		return 0;
	}

	/* else */
	return usage(1);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
		return usage(0);

	if (strcmp(argv[1], "version") == 0)
		return cpt_version(argc - 1, argv + 1);
	else if (strcmp(argv[1], "caps") == 0)
		return cpt_caps(argc - 1, argv + 1);
	else if (strcmp(argv[1], "help") == 0)
		return usage(0);

	return usage(1);
}
